### PR TITLE
feat: use full jitter exp backoff policy in the generator

### PR DESCRIPTION
**Pull request by alevenberg**
Fixes issue #8755

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)

### CODE REVIEW

**Review by coryan — APPROVED** [review]

**Review by dbolduc — CHANGES_REQUESTED** [review]
> Reviewed 3 of 233 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)* status: 3 of 233 files reviewed, 2 unresolved discussions (waiting on @alevenberg and @coryan)

___
*[`generator/internal/option_defaults_generator.cc` line 166 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#-NWiF5E16xIwkTfORy4y:-NWiF5E16xIwkTfORy4z:boji137) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1f24eecf935bb7e1d8898a0ecc7bd84857275038/generator/internal/option_defaults_generator.cc#L166)):*
> ```cpp
>     "            $product_namespace$::$service_name$BackoffPolicyOption::Type>(\n"
>     "            options.get<$product_namespace$::$retry_policy_name$Option>()->clone(),\n"
>     "            options.get<$product_namespace$::$service_name$BackoffPolicyOption>()->clone())\n"
> ```

We should change this back to the old value.

**Review by alevenberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)* status: 3 of 233 files reviewed, 1 unresolved discussion (waiting on @coryan and @dbolduc)

___
*[`generator/internal/option_defaults_generator.cc` line 166 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#-NWiF5E16xIwkTfORy4y:-NWiFbri-MM50vJZL_n_:b-lbut5x) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1f24eecf935bb7e1d8898a0ecc7bd84857275038/generator/internal/option_defaults_generator.cc#L166)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

We should change this back to the old value.
</blockquote></details>

What do you mean?

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)* status: 3 of 233 files reviewed, 1 unresolved discussion (waiting on @alevenberg and @coryan)

___
*[`generator/internal/option_defaults_generator.cc` line 166 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#-NWiF5E16xIwkTfORy4y:-NWiLDNX2LFcGM_bLDxS:b-2abla3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1f24eecf935bb7e1d8898a0ecc7bd84857275038/generator/internal/option_defaults_generator.cc#L166)):*
<details><summary><i>Previously, alevenberg (Anna Levenberg) wrote…</i></summary><blockquote>

What do you mean?
</blockquote></details>

We should use `ExponentialBackoffPolicy(std::chrono::seconds(1), std::chrono::minutes(5), kBackoffScaling)` instead of cloning the `*BackoffPolicyOption`


(Because full jitter is good for individual retries to address the thundering herd problem. But it is not best for polling LROs. We do not want to be waking up background threads after 0 seconds to check the status of the operation.)

**Review by alevenberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)* status: 2 of 233 files reviewed, 1 unresolved discussion (waiting on @coryan and @dbolduc)

___
*[`generator/internal/option_defaults_generator.cc` line 166 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#-NWiF5E16xIwkTfORy4y:-NWiY8TK4lSFQ8yG-LIP:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1f24eecf935bb7e1d8898a0ecc7bd84857275038/generator/internal/option_defaults_generator.cc#L166)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

We should use `ExponentialBackoffPolicy(std::chrono::seconds(1), std::chrono::minutes(5), kBackoffScaling)` instead of cloning the `*BackoffPolicyOption`


(Because full jitter is good for individual retries to address the thundering herd problem. But it is not best for polling LROs. We do not want to be waking up background threads after 0 seconds to check the status of the operation.) 
</blockquote></details>

Done.

File: generator/internal/option_defaults_generator.cc [hunk 1]
```diff
@@ -158,13 +158,9 @@ Status OptionDefaultsGenerator::GenerateCc() {
     "  }\n"},
    {[this]{return HasLongrunningMethod();},
     "  if (!options.has<$product_namespace$::$service_name$PollingPolicyOption>()) {\n"
-    "    options.set<$product_namespace$::$service_name$PollingPolicyOption>(\n"
-    "        GenericPollingPolicy<\n"
-    "            $product_namespace$::$retry_policy_name$Option::Type,\n"
-    "            $product_namespace$::$service_name$BackoffPolicyOption::Type>(\n"
-    "            options.get<$product_namespace$::$retry_policy_name$Option>()->clone(),\n"
-    "            options.get<$product_namespace$::$service_name$BackoffPolicyOption>()->clone())\n"
```
Comment (dbolduc) [inline generator/internal/option_defaults_generator.cc:166]: This is the only line we want to change.

File: generator/internal/option_defaults_generator.cc [hunk 2]
```diff
@@ -158,13 +158,9 @@ Status OptionDefaultsGenerator::GenerateCc() {
...
-    "            .clone());\n"
+    "    options.set<$product_namespace$::$service_name$BackoffPolicyOption>(\n"
```
Comment (dbolduc) [inline generator/internal/option_defaults_generator.cc:161]: we want to set the PollingPolicyOption

**Review by dbolduc — COMMENTED** [review]
It might be worth adding tests to ...

- https://github.com/googleapis/google-cloud-cpp/blob/main/generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc
- https://github.com/googleapis/google-cloud-cpp/blob/main/generator/integration_tests/tests/golden_thing_admin_option_defaults_test.cc

... to verify that the retry / backoff / polling policy options have values. (We wouldn't need to check what their value is.)

File: generator/internal/option_defaults_generator.cc (hunk 2 above)
Comment (alevenberg) [reply generator/internal/option_defaults_generator.cc:161]: Fixed

Comment (dbolduc) [reply generator/internal/option_defaults_generator.cc:161]: `git push` ?

File: generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc [hunk 3]
```diff
@@ -25,6 +26,10 @@ namespace golden_v1_internal {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace {
 
+using ::google::cloud::golden_v1::v2_11::GoldenKitchenSinkBackoffPolicyOption;
+using ::google::cloud::golden_v1::v2_11::
+    GoldenKitchenSinkConnectionIdempotencyPolicyOption;
+using ::google::cloud::golden_v1::v2_11::GoldenKitchenSinkRetryPolicyOption;
```
Comment (dbolduc) [inline generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc:32]: we should not be using the [inline namespace](https://en.cppreference.com/w/cpp/language/namespace#Inline_namespaces), `v2_11`. Inline namespaces are only used by libraries/applications that need to use two versions of `google-cloud-cpp` at the same time. (which is hopefully rare).

```cc
using ::google::cloud::golden_v1::GoldenKitchenSinkBackoffPolicyOption;
```

File: generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc [hunk 4]
```diff
@@ -80,6 +85,26 @@ TEST(GoldenKitchenSinkDefaultOptions, UserProjectOptionAndEnvVar) {
   EXPECT_EQ("test-project", updated_options.get<UserProjectOption>());
 }
 
+TEST(GoldenKitchenSinkDefaultOptions, RetryPolicySet) {
+  auto options = Options{};
+  auto updated_options = GoldenKitchenSinkDefaultOptions(options);
+  EXPECT_EQ(true, updated_options.has<GoldenKitchenSinkRetryPolicyOption>());
```
Comment (dbolduc) [inline generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc:91]: nit: here and below, we always say:

```cc
EXPECT_TRUE(updated_options.has<GoldenKitchenSinkRetryPolicyOption>());
```

File: generator/integration_tests/tests/golden_thing_admin_option_defaults_test.cc [hunk 5]
```diff
@@ -25,6 +26,9 @@ namespace golden_v1_internal {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace {
 
+using ::google::cloud::golden_v1::GoldenThingAdminBackoffPolicyOption;
```
Comment (dbolduc) [inline generator/integration_tests/tests/golden_thing_admin_option_defaults_test.cc:29]: these are right :+1:

**Review by dbolduc — APPROVED** [review]
The inline namespace thing needs to get fixed, but I trust you can make the change.

**Review by dbolduc — APPROVED** [review]

**Review by alevenberg — COMMENTED** [review]
Done

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#-:-NWnGymF7500h_1eQaHO:bb6un1j)* status: 2 of 235 files reviewed, 5 unresolved discussions (waiting on @coryan and @dbolduc)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#gh-1210845141:-NWnH04S9Y2pDuaU8LEK:bb6un1j) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

This is the only line we want to change.
</blockquote></details>

Done

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#gh-1210845260:-NWnH11x8V0_kZQHduv6:bh9f965) (no related file):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

`git push` ?
</blockquote></details>

Whoops. Done

___
___
*[`generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc` line 32 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#gh-1212124010:-NWnH4yGC8Hht9TLSADZ:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/76ecfbb3aa9960bf6b5bfa9664e5adb77362a14c/generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc#L32)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

we should not be using the [inline namespace](https://en.cppreference.com/w/cpp/language/namespace#Inline_namespaces), `v2_11`. Inline namespaces are only used by libraries/applications that need to use two versions of `google-cloud-cpp` at the same time. (which is hopefully rare).

```cc
using ::google::cloud::golden_v1::GoldenKitchenSinkBackoffPolicyOption;
```
</blockquote></details>

Done
___
*[`generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc` line 91 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#gh-1212124448:-NWnH5pgBcFra-2susnJ:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/76ecfbb3aa9960bf6b5bfa9664e5adb77362a14c/generator/integration_tests/tests/golden_kitchen_sink_option_defaults_test.cc#L91)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: here and below, we always say:

```cc
EXPECT_TRUE(updated_options.has<GoldenKitchenSinkRetryPolicyOption>());
```
</blockquote></details>

Done
___
*[`generator/integration_tests/tests/golden_thing_admin_option_defaults_test.cc` line 29 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748#gh-1212127021:-NWnH6l8ALKzPEMlDft8:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/76ecfbb3aa9960bf6b5bfa9664e5adb77362a14c/generator/integration_tests/tests/golden_thing_admin_option_defaults_test.cc#L29)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

these are right :+1: 
</blockquote></details>

Done.

**Review by alevenberg — COMMENTED** [review]
> Dismissed @dbolduc from 5 discussions.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)* status: 2 of 235 files reviewed, all discussions resolved (waiting on @coryan and @dbolduc)

**Review by coryan — APPROVED** [review]
> Reviewed 122 of 233 files at r1, 111 of 113 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11748)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @alevenberg)

**Pull request merged**